src/Associativity.cpp

```diff
@@ -46,195 +59,385 @@ class ConvertSelfRef : public IRMutator {
         // anything since it's going to be replaced anyway, but it's not
         // commutative
-        op.op = y;
-        op.identity = make_const(t, 0);
-        op.x = {"", Expr()};
-        op.y = {op_y, e};
-        return {true, false};
+        assoc_op.pattern.ops[0] = Variable::make(t, op_y_names[0]);
+        assoc_op.pattern.identities[0] = make_const(t, 0);
+        assoc_op.pattern.is_commutative = false;
+        assoc_op.xs[0] = {"", Expr()};
+        assoc_op.ys[0] = {op_y_names[0], exprs[0]};
+        return true;
     }
+    return find_match(get_ops_table(exprs), op_x_names, op_y_names,
+                      x_parts, exprs, assoc_op);
+}
 
-    if (const Add *a = e.as<Add>()) {
-        op.op = x + y;
-        op.identity = make_const(t, 0);
-        return visit_associative_binary_op<Add>(op_x, op_y, x_part, a->a, a->b, op);
-    } else if (const Sub *s = e.as<Sub>()) {
-        op.op = x + y;
-        op.identity = make_const(t, 0);
-        return visit_associative_binary_op<Sub>(op_x, op_y, x_part, s->a, s->b, op);
-    } else if (const Mul *m = e.as<Mul>()) {
-        op.op = x * y;
-        op.identity = make_const(t, 1);
-        return visit_associative_binary_op<Mul>(op_x, op_y, x_part, m->a, m->b, op);
-    } else if (const Min *m = e.as<Min>()) {
-        op.op = Min::make(x, y);
-        op.identity = t.max();
-        return visit_associative_binary_op<Min>(op_x, op_y, x_part, m->a, m->b, op);
-    } else if (const Max *m = e.as<Max>()) {
-        op.op = Max::make(x, y);
-        op.identity = t.min();
-        return visit_associative_binary_op<Max>(op_x, op_y, x_part, m->a, m->b, op);
-    } else if (const And *a = e.as<And>()) {
-        op.op = And::make(x, y);
-        op.identity = make_const(t, 1);
-        return visit_associative_binary_op<And>(op_x, op_y, x_part, a->a, a->b, op);
-    } else if (const Or *o = e.as<Or>()) {
-        op.op = Or::make(x, y);
-        op.identity = make_const(t, 0);
-        return visit_associative_binary_op<Or>(op_x, op_y, x_part, o->a, o->b, op);
-    } else if (e.as<Let>()) {
-        internal_error << "Let should have been substituted before calling this function\n";
-    } else {
-        debug(4) << "Can't prove associativity of " << e << "\n";
-        return {false, false};
+void add_transitive_dependencies(vector<set<int>> &dependencies) {
+    // TODO(psuriana): there might be a better way to find all the transitive dependencies
+    bool change = true;
+    while (change) {
+        change = false;
+        for (size_t i = 0; i < dependencies.size(); ++i) {
+            for (size_t j = 0; j < dependencies.size(); ++j) {
+                if (i == j) {
+                    continue;
+                }
+                if (dependencies[i].find(j) != dependencies[i].end()) {
+                    for (const auto &idx : dependencies[j]) {
+                        if (dependencies[i].find(idx) == dependencies[i].end()) {
```

Comment:
I need to update the "change" flag so that I know when to break from the while loop